torch/csrc/distributed/c10d/init.cpp

```diff
@@ -992,6 +1010,36 @@ that adds a prefix to each key inserted to the store.
           "get_future",
           [](::c10d::ProcessGroup::Work& work)
               -> std::shared_ptr<jit::PythonFutureWrapper> {
+#ifdef USE_C10D_NCCL
+            auto fut = dynamic_cast<::c10d::ProcessGroupNCCL::FutureNCCL*>(
+                work.getFuture().get());
+            // By default record_stream_cb_ of the future is not set, and here
+            // we assume the type of this future value is or can be casted into
+            // a tensor vector that only has exactly one tensor.
+            if (!fut->hasRecordStreamCallback()) {
+              auto record_stream_cb = [](const at::IValue& value,
+                                         at::cuda::CUDAStream cuda_stream) {
+                if (value.isTensorList() || value.isPyObject()) {
+                  pybind11::gil_scoped_acquire gil;
+                  std::vector<at::Tensor> tensors;
+                  if (value.isTensorList()) {
+                    tensors = std::move(value.toTensorVector());
+                  } else {
+                    py::object obj = torch::jit::toPyObject(value);
```

Comment:
We probably need to acquire GIL only for this codepath? Can we move `pybind11::gil_scoped_acquire gil;` here?